File: packages/redact/src/redact.ts
import type { TraceEvent } from "@runback/schema";
import {
  RULES,
  SENSITIVE_KEYS,
  applyRule,
  BASE64_CANDIDATE,
  decodeBase64Utf8,
  type RedactionRule,
  type RedactionTier,
} from "./patterns";

export interface RedactOptions {
  /** Which built-in tier to use. `standard` (default) is high-confidence only. */
  preset?: RedactionTier;
  /** Replace the built-in rule set entirely. */
  rules?: RedactionRule[];
  /** Add rules on top of the preset. */
  customPatterns?: Array<{ name: string; regex: RegExp; replace?: (m: string) => string }>;
  /** Extra object-key names whose values are redacted wholesale. */
  redactKeys?: string[];
  /** Key names to never touch (takes precedence over everything). */
  allowKeys?: string[];
  /** Full escape hatch. Return a replacement, or `undefined` to fall through to defaults. */
  redactor?: (value: unknown, key: string | null) => unknown;
  /** Called once per event with how many values were redacted. */
  onRedact?: (info: { redactions: number }) => void;
  /** Max recursion depth (cycle/blowup guard). */
  maxDepth?: number;
  /** Max characters kept per string field before it's truncated with a
   *  marker — a size guard against one huge tool output or pasted document
   *  bloating every ingest payload, independent of redaction. Applied BEFORE
   *  redaction's own regex passes run, so a huge field can't also make
   *  redaction itself slow (a long run of base64-alphabet characters can
   *  make the base64-candidate scan pathologically slow — measured 3+
   *  seconds on a single 60,000-char field pre-truncation). Default 50,000
   *  (~50KB) — generous for real content, small next to what an unbounded
   *  blob could be. */
  maxStringLength?: number;
}

/** One redaction — never the matched value itself, only what kind and where. */
export interface RedactionLogEntry {
  /** Rule/pattern name (e.g. "email", "ssn"), or "key:<name>" for a whole-value key match. */
  rule: string;
  /** JSON path where it occurred, e.g. "request.messages[0].content" — null at the root. */
  path: string | null;
  ts: string;
}

export interface Redactor {
  /** Redact an arbitrary JSON value. */
  redactValue(value: unknown): unknown;
  /** Redact the content-bearing fields of a trace event; envelope/metrics preserved. */
  redactEvent(event: TraceEvent): TraceEvent;
  /** Total values redacted across this redactor's lifetime. */
  total(): number;
  /** Itemized log of redactions across this redactor's lifetime, capped at 500 entries. */
  log(): RedactionLogEntry[];
  /** True once the log hit its cap — total() stays accurate even after this. */
  logTruncated(): boolean;
  /** How many string fields have been size-truncated (maxStringLength) across this redactor's lifetime — independent of redaction count. */
  truncatedFields(): number;
}

type Normalized = boolean | RedactionTier | RedactOptions | undefined;

/** Accepts `true`/`'standard'`/`'strict'`/options object/`false`. Returns null to disable. */
export function createRedactor(input?: Normalized): Redactor | null {
  if (input === false || input === undefined) return null;
  const opts: RedactOptions =
    input === true ? {} : typeof input === "string" ? { preset: input } : input;

  const tier = opts.preset ?? "standard";
  const baseRules = (opts.rules ?? RULES).filter(
    (r) => tier === "strict" || r.tier === "standard"
  );
  const customRules: RedactionRule[] = (opts.customPatterns ?? []).map((p) => ({
    name: p.name,
    regex: p.regex,
    tier: "standard",
    replace: p.replace,
  }));
  const activeRules = [...baseRules, ...customRules];

  const sensitive = new Set(
    [...SENSITIVE_KEYS, ...(opts.redactKeys ?? [])].map((k) => k.toLowerCase())
  );
  const allow = new Set((opts.allowKeys ?? []).map((k) => k.toLowerCase()));
  const maxDepth = opts.maxDepth ?? 24;
  const maxStringLength = opts.maxStringLength ?? 50_000;
  let lifetime = 0;
  let fieldsTruncated = 0;
  const MAX_LOG = 500;
  const redactionLog: RedactionLogEntry[] = [];
  let truncated = false;

  function record(rule: string, path: string | null) {
    if (redactionLog.length < MAX_LOG) redactionLog.push({ rule, path, ts: new Date().toISOString() });
    else truncated = true;
  }

  /** Size guard, independent of redaction — see RedactOptions.maxStringLength.
   *  Deliberately NOT recorded into the redaction log/total(): a truncated
   *  field wasn't found sensitive, and folding it into "N values redacted"
   *  would misrepresent why it changed to anyone reading that count. */
  function capLength(s: string, _path: string | null): string {
    if (s.length <= maxStringLength) return s;
    fieldsTruncated++;
    const overBy = s.length - maxStringLength;
    return `${s.slice(0, maxStringLength)}…[truncated: ${overBy} more chars]`;
  }

  function redactString(s: string, counter: { n: number }, path: string | null): string {
    let out = s;
    for (const rule of activeRules) {
      const { out: next, count } = applyRule(rule, out, () => record(rule.name, path));
      out = next;
      counter.n += count;
    }
    return redactBase64Secrets(out, counter, path);
  }

  /**
   * A secret pasted through base64 doesn't match any plaintext rule above —
   * the encoded text just doesn't look like an API key or a JWT. This finds
   * base64-shaped substrings, decodes each, and re-checks the DECODED text
   * against every active rule; a real hit redacts the original encoded
   * substring wholesale. Legitimate base64 (an image blob, a content hash)
   * decodes to bytes that don't match anything and is left untouched.
   */
  function redactBase64Secrets(s: string, counter: { n: number }, path: string | null): string {
    return s.replace(BASE64_CANDIDATE, (candidate) => {
      const decoded = decodeBase64Utf8(candidate);
      if (decoded == null) return candidate;
      for (const rule of activeRules) {
        const { count } = applyRule(rule, decoded);
        if (count > 0) {
          counter.n++;
          record(`base64:${rule.name}`, path);
          return "[redacted:base64]";
        }
      }
      return candidate;
    });
  }

  function walk(
    value: unknown,
    key: string | null,
    depth: number,
    seen: WeakSet<object>,
    counter: { n: number },
    path: string | null
  ): unknown {
    // custom escape hatch first
    if (opts.redactor) {
      const r = opts.redactor(value, key);
      if (r !== undefined) {
        if (r !== value) { counter.n++; record("custom", path); }
        return r;
      }
    }

    if (key !== null) {
      const lk = key.toLowerCase();
      if (allow.has(lk)) return value;
      if (sensitive.has(lk) && value != null) {
        counter.n++;
        record(`key:${key}`, path);
        return `[redacted:key:${key}]`;
      }
    }

    if (value == null) return value;
    // Cap BEFORE redacting, not after: redaction runs several regex passes
    // (including a base64-candidate scan) over the whole string, and a long
    // run of a base64-alphabet character can make that scan pathologically
    // slow — measured 3+ seconds on a single 60,000-char field. Truncating
    // first bounds redaction's own cost to at most maxStringLength, and
    // whatever gets cut is never sent anywhere, so there's no leak risk in
    // skipping it.
    if (typeof value === "string") return redactString(capLength(value, path), counter, path);
    if (typeof value === "number" || typeof value === "boolean") return value;
    if (depth >= maxDepth) return value;

    if (Array.isArray(value)) {
      if (seen.has(value)) return value;
      seen.add(value);
      return value.map((v, i) => walk(v, null, depth + 1, seen, counter, `${path ?? ""}[${i}]`));
    }
    if (typeof value === "object") {
      if (seen.has(value as object)) return value;
      seen.add(value as object);
      const out: Record<string, unknown> = {};
      for (const [k, v] of Object.entries(value as Record<string, unknown>)) {
        out[k] = walk(v, k, depth + 1, seen, counter, path ? `${path}.${k}` : k);
      }
      return out;
    }
    return value;
  }

  function redactValue(value: unknown): unknown {
    const counter = { n: 0 };
    const out = walk(value, null, 0, new WeakSet(), counter, null);
    lifetime += counter.n;
    return out;
  }

  function redactEvent(event: TraceEvent): TraceEvent {
    const counter = { n: 0 };
    const seen = new WeakSet<object>();
    const w = (v: unknown, k: string | null = null, path: string | null = k) => walk(v, k, 0, seen, counter, path);

    // Default to the event itself: a type with no case below is returned as-is,
    // never silently dropped. (The `env` case was once missing, which dropped env
    // events whenever redaction was on — this is the runtime guard against a repeat.)
    let out: TraceEvent = event;
    switch (event.type) {
      case "llm":
        out = {
          ...event,
          request: {
            ...event.request,
            system: event.request.system == null ? null : (w(event.request.system, "system", "request.system") as string),
            messages: w(event.request.messages, "messages", "request.messages") as typeof event.request.messages,
            tools: w(event.request.tools, "tools", "request.tools") as typeof event.request.tools,
          },
          response: {
            ...event.response,
            text: event.response.text == null ? null : (w(event.response.text, "text", "response.text") as string),
            reasoning:
              event.response.reasoning == null ? null : (w(event.response.reasoning, "reasoning", "response.reasoning") as string),
            tool_calls: w(event.response.tool_calls, "tool_calls", "response.tool_calls") as typeof event.response.tool_calls,
          },
          error: event.error ? (w(event.error, "error") as typeof event.error) : null,
        };
        break;
      case "tool":
        out = {
          ...event,
          input: w(event.input, "input"),
          output: event.output == null ? null : w(event.output, "output"),
          error: event.error ? (w(event.error, "error") as typeof event.error) : null,
        };
        break;
      case "reasoning":
        out = { ...event, text: w(event.text, "text") as string };
        break;
      case "env":
        // Env reads are oracle entries too: a captured `fetch` output carries a
        // response body that can hold secrets/PII, so scrub it. Redaction runs
        // BEFORE the event is hash-chained (in the collector), so the recorded and
        // replayed values stay identical and the digest is computed over the
        // scrubbed value — consistent end to end. Primitive reads (now/random/
        // uuid/date) walk to themselves, leaving the digest untouched.
        out = { ...event, output: event.output == null ? null : w(event.output, "output") };
        break;
      case "run":
        out = {
          ...event,
          input: event.input == null ? null : w(event.input, "input"),
          output: event.output == null ? null : w(event.output, "output"),
          metadata: w(event.metadata, "metadata") as Record<string, unknown>,
          error: event.error ? (w(event.error, "error") as typeof event.error) : null,
        };
        break;
    }
    lifetime += counter.n;
    opts.onRedact?.({ redactions: counter.n });
    return out;
  }

  return {
    redactValue,
    redactEvent,
    total: () => lifetime,
    log: () => redactionLog,
    logTruncated: () => truncated,
    truncatedFields: () => fieldsTruncated,
  };
}

File: packages/sdk-python/runback/redact.py
"""
Faithful port of packages/redact/src/redact.ts. Same defaults, same walk
order, same per-event-type field list — see test_redact.py for fixture
parity against the TypeScript implementation.
"""

from __future__ import annotations

import re
from dataclasses import dataclass
from datetime import datetime, timezone
from typing import Any, Callable, Optional, Union

from .patterns import (
    BASE64_CANDIDATE,
    RULES,
    SENSITIVE_KEYS,
    RedactionRule,
    apply_rule,
    decode_base64_utf8,
)

RedactionTier = str  # "standard" | "strict"


@dataclass
class RedactOptions:
    # Which built-in tier to use. "standard" (default) is high-confidence only.
    preset: Optional[RedactionTier] = None
    # Replace the built-in rule set entirely.
    rules: Optional[list[RedactionRule]] = None
    # Add rules on top of the preset: (name, compiled_regex, optional replace fn).
    custom_patterns: Optional[list[tuple[str, "re.Pattern[str]", Optional[Callable[[str], str]]]]] = None
    # Extra object-key names whose values are redacted wholesale.
    redact_keys: Optional[list[str]] = None
    # Key names to never touch (takes precedence over everything).
    allow_keys: Optional[list[str]] = None
    # Full escape hatch. Return a replacement, or None to fall through to defaults.
    redactor: Optional[Callable[[Any, Optional[str]], Any]] = None
    # Called once per event with how many values were redacted.
    on_redact: Optional[Callable[[int], None]] = None
    # Max recursion depth (cycle/blowup guard).
    max_depth: int = 24


@dataclass
class RedactionLogEntry:
    rule: str
    path: Optional[str]
    ts: str


# What create_redactor() accepts: True / "standard" / "strict" / RedactOptions / False / None.
RedactorInput = Union[bool, RedactionTier, RedactOptions, None]


class Redactor:
    """Returned by create_redactor(). Mirrors the TS Redactor interface's four methods."""

    def __init__(self, opts: RedactOptions) -> None:
        self._opts = opts
        tier = opts.preset or "standard"
        base_rules = [r for r in (opts.rules or RULES) if tier == "strict" or r.tier == "standard"]
        custom_rules = [
            RedactionRule(name=name, regex=regex, tier="standard", replace=replace)
            for (name, regex, replace) in (opts.custom_patterns or [])
        ]
        self._active_rules = [*base_rules, *custom_rules]
        self._sensitive = {k.lower() for k in [*SENSITIVE_KEYS, *(opts.redact_keys or [])]}
        self._allow = {k.lower() for k in (opts.allow_keys or [])}
        self._max_depth = opts.max_depth
        self._lifetime = 0
        self._max_log = 500
        self._log: list[RedactionLogEntry] = []
        self._truncated = False

    def _record(self, rule: str, path: Optional[str]) -> None:
        if len(self._log) < self._max_log:
            self._log.append(RedactionLogEntry(rule=rule, path=path, ts=datetime.now(timezone.utc).isoformat()))
        else:
            self._truncated = True

    def _redact_string(self, s: str, counter: list[int], path: Optional[str]) -> str:
        out = s
        for rule in self._active_rules:
            out, count = apply_rule(rule, out, lambda r=rule: self._record(r.name, path))
            counter[0] += count
        return self._redact_base64_secrets(out, counter, path)

    def _redact_base64_secrets(self, s: str, counter: list[int], path: Optional[str]) -> str:
        def _sub(m: "re.Match[str]") -> str:
            candidate = m.group(0)
            decoded = decode_base64_utf8(candidate)
            if decoded is None:
                return candidate
            for rule in self._active_rules:
                _, count = apply_rule(rule, decoded)
                if count > 0:
                    counter[0] += 1
                    self._record(f"base64:{rule.name}", path)
                    return "[redacted:base64]"
            return candidate

        return BASE64_CANDIDATE.sub(_sub, s)

    def _walk(
        self,
        value: Any,
        key: Optional[str],
        depth: int,
        seen: set[int],
        counter: list[int],
        path: Optional[str],
    ) -> Any:
        if self._opts.redactor:
            r = self._opts.redactor(value, key)
            if r is not None:
                if r != value:
                    counter[0] += 1
                    self._record("custom", path)
                return r

        if key is not None:
            lk = key.lower()
            if lk in self._allow:
                return value
            if lk in self._sensitive and value is not None:
                counter[0] += 1
                self._record(f"key:{key}", path)
                return f"[redacted:key:{key}]"

        if value is None:
            return value
        if isinstance(value, str):
            return self._redact_string(value, counter, path)
        if isinstance(value, bool) or isinstance(value, (int, float)):
            return value
        if depth >= self._max_depth:
            return value

        if isinstance(value, list):
            oid = id(value)
            if oid in seen:
                return value
            seen = seen | {oid}
            return [
                self._walk(v, None, depth + 1, seen, counter, f"{path or ''}[{i}]") for i, v in enumerate(value)
            ]
        if isinstance(value, dict):
            oid = id(value)
            if oid in seen:
                return value
            seen = seen | {oid}
            return {
                k: self._walk(v, k, depth + 1, seen, counter, f"{path}.{k}" if path else k)
                for k, v in value.items()
            }
        return value

    def redact_value(self, value: Any) -> Any:
        counter = [0]
        out = self._walk(value, None, 0, set(), counter, None)
        self._lifetime += counter[0]
        return out

    def redact_event(self, event: dict) -> dict:
        """
        Redact the content-bearing fields of a trace event; envelope/metrics
        preserved. Operates on a plain dict matching the wire schema (this
        SDK has no TS-style discriminated-union type system) — same
        per-type field list as redact.ts's switch, including the "unknown
        type passes through unmodified" fallback.
        """
        counter = [0]
        seen: set[int] = set()

        def w(v: Any, k: Optional[str] = None, path: Optional[str] = None) -> Any:
            return self._walk(v, k, 0, seen, counter, path if path is not None else k)

        event_type = event.get("type")
        out = dict(event)

        if event_type == "llm":
            request = dict(event["request"])
            request["system"] = None if request.get("system") is None else w(request["system"], "system", "request.system")
            request["messages"] = w(request.get("messages"), "messages", "request.messages")
            request["tools"] = w(request.get("tools"), "tools", "request.tools")
            response = dict(event["response"])
            response["text"] = None if response.get("text") is None else w(response["text"], "text", "response.text")
            response["reasoning"] = (
                None if response.get("reasoning") is None else w(response["reasoning"], "reasoning", "response.reasoning")
            )
            response["tool_calls"] = w(response.get("tool_calls"), "tool_calls", "response.tool_calls")
            out["request"] = request
            out["response"] = response
            out["error"] = w(event["error"], "error") if event.get("error") else None
        elif event_type == "tool":
            out["input"] = w(event.get("input"), "input")
            out["output"] = None if event.get("output") is None else w(event["output"], "output")
            out["error"] = w(event["error"], "error") if event.get("error") else None
        elif event_type == "reasoning":
            out["text"] = w(event.get("text"), "text")
        elif event_type == "env":
            # Env reads are oracle entries too — a captured fetch output can
            # carry secrets/PII, so scrub it. Primitive reads (now/random/
            # uuid/date) walk to themselves, leaving the digest untouched.
            out["output"] = None if event.get("output") is None else w(event["output"], "output")
        elif event_type == "run":
            out["input"] = None if event.get("input") is None else w(event["input"], "input")
            out["output"] = None if event.get("output") is None else w(event["output"], "output")
            out["metadata"] = w(event.get("metadata"), "metadata")
            out["error"] = w(event["error"], "error") if event.get("error") else None
        # else: unknown type, returned as-is — never silently dropped.

        self._lifetime += counter[0]
        if self._opts.on_redact:
            self._opts.on_redact(counter[0])
        return out

    def total(self) -> int:
        return self._lifetime

    def log(self) -> list[RedactionLogEntry]:
        return self._log

    def log_truncated(self) -> bool:
        return self._truncated


def create_redactor(input: RedactorInput = True) -> Optional[Redactor]:
    """Accepts True/"standard"/"strict"/RedactOptions/False/None. Returns None to disable."""
    if input is False or input is None:
        return None
    if input is True:
        opts = RedactOptions()
    elif isinstance(input, str):
        opts = RedactOptions(preset=input)
    else:
        opts = input
    return Redactor(opts)
